### DISCUSSION TITLE
Grouping dependencies

### DISCUSSION THREAD

**Discussion by HonkingGoose** [renovatebot/renovate#17266] [Request Help] [answered]
## Question

I'm looking for a good way to limit the amount of PRs we get for the Octoclairvoyant repository. The big goal is to be up-to-date with the framework stack at all times, so any security update is easy to apply. Then for all other things we want to get updates once a month, in grouped PRs.

The problem I'm having is that it's hard to decide what to group together, so that things won't break as often.

## Proposed grouping

Would grouping things like this make sense?

| Grouping | Security critical | Update frequency |
| --- | --- | --- |
| Next.js, React (Framework stack) | yes | daily |
| ESLint, Prettier, ESLint plugins | not likely | monthly |
| Jest, Cypress, and testing utils | not likely |  monthly |
| Chakra UI, emotion, and Framer Motion | not likely | monthly |
| React Query + Octokit | not likely | monthly |
| Rehype + Remark ecosystem | not likely | monthly |
| TypeScript, TS types, and type-fest | not likely | monthly |
| lint-staged and husky | not likely | monthly |

## Current status

<details><summary>Current package.json dependencies and devDependencies</summary>

```json
	"dependencies": {
		"@chakra-ui/react": "2.2.8",
		"@emotion/react": "11.10.0",
		"@emotion/styled": "11.10.0",
		"@octokit/auth-callback": "3.0.0",
		"@octokit/rest": "19.0.3",
		"@sentry/nextjs": "7.10.0",
		"@tanstack/react-query": "4.1.0",
		"@tanstack/react-query-devtools": "4.0.10",
		"downshift": "6.1.9",
		"framer-motion": "7.1.0",
		"lodash-es": "4.17.21",
		"next": "12.2.5",
		"next-seo": "5.5.0",
		"nookies": "2.5.2",
		"react": "18.2.0",
		"react-dom": "18.2.0",
		"react-icons": "4.4.0",
		"rehype-highlight": "5.0.2",
		"rehype-react": "7.1.1",
		"remark-gfm": "3.0.1",
		"remark-github": "11.2.4",
		"remark-parse": "10.0.1",
		"remark-rehype": "10.1.0",
		"remark-stringify": "10.0.2",
		"semver": "7.3.7",
		"unified": "10.1.2"
	},
	"devDependencies": {
		"@chakra-ui/cli": "2.1.6",
		"@chakra-ui/styled-system": "2.2.7",
		"@chakra-ui/theme-tools": "2.0.9",
		"@chakra-ui/utils": "2.0.8",
		"@jest/globals": "28.1.3",
		"@octokit/openapi-types": "12.11.0",
		"@testing-library/cypress": "8.0.3",
		"@types/happo-cypress": "3.0.1",
		"@types/lodash": "4.14.182",
		"@types/mdast": "3.0.10",
		"@types/node": "16.11.47",
		"@types/react": "18.0.17",
		"@types/react-dom": "18.0.6",
		"@types/semver": "7.3.12",
		"@types/unist": "2.0.6",
		"@typescript-eslint/eslint-plugin": "5.33.0",
		"@typescript-eslint/parser": "5.33.0",
		"cypress": "10.4.0",
		"eslint": "8.22.0",
		"eslint-config-next": "12.2.5",
		"eslint-config-prettier": "8.5.0",
		"eslint-plugin-cypress": "2.12.1",
		"eslint-plugin-jest": "26.8.3",
		"eslint-plugin-jest-formatting": "3.1.0",
		"eslint-plugin-unicorn": "43.0.2",
		"happo-cypress": "3.0.1",
		"happo-e2e": "1.2.0",
		"happo.io": "7.3.0",
		"husky": "8.0.1",
		"is-ci": "3.0.1",
		"jest": "28.1.3",
		"lint-staged": "13.0.3",
		"prettier": "2.7.1",
		"start-server-and-test": "1.14.0",
		"type-fest": "2.18.0",
		"typescript": "4.7.4"
	},
```

</details>

<details><summary>Current renovate.json config</summary>

```json
{
  "extends": [
    "config:base",
    "helpers:pinGitHubActionDigests",
    ":automergeTypes"
  ],
  "prCreation": "not-pending",
  "lockFileMaintenance": {
    "enabled": true,
    "automerge": true
  },
  "internalChecksFilter": "strict",
  "packageRules": [
    {
      "description": "Wait 3 days before creating a npm update PR",
      "matchDatasources": ["npm"],
      "stabilityDays": 3
    },
    {
      "description": "Automerge ESLint and Prettier updates",
      "matchDepTypes": ["devDependencies"],
      "matchPackagePatterns": ["eslint", "prettier"],
      "automerge": true
    },
    {
      "description": "Get pnpm updates once a month",
      "matchDepTypes": ["packageManager"],
      "matchPackageNames": ["pnpm"],
      "extends": ["schedule:monthly"]
    }
  ]
}
```

</details>

**Comment by rarkins** [comment, marked as answer]
Consider grouping dev dependencies monthly for all non-major?

**Reply by HonkingGoose** [reply]
### `packageRules` to group dev dependencies once a month (non-major)

I think this would work then?

```json
{
  "packageRules": [
    {
      "description": ["Group all non-major yarn development dependencies together, get updates once a month"],
      "matchManagers": ["npm"],
      "matchDepTypes": ["devDependencies"],
      "matchUpdateTypes": ["minor", "patch"],
      "groupName": "Group all non-major dev dependencies",
      "extends": ["schedule:monthly"]
    }
  ]
}
```

### Ordering the rules in `packageRules`

The `packageRules` config option docs say:

> Important to know: Renovate will evaluate all `packageRules` and not stop once it gets a first match. You should order your `packageRules` in order of importance so that later rules can override settings from earlier rules if needed.

Hmm, so does that mean the `packageRules` rules order should be:

1. Group non-major dev updates
2. The "get pnpm updates once a month" part
3. Automerge ESLint and Prettier updates
4. The 3 day waiting period for npm packages

Or should I order it differently?

**Reply by rarkins** [reply]
I'm not sure you need 3 day waiting period last. If you mean setting stabilityDays then you don't need it to be in any particular order. Also you may want to set internalChecksFilter=strict

**Reply by HonkingGoose** [reply]
Thank you for your help! I'll go try out the new config, and see what happens next. 😄
